state/protocol/inmem/epoch.go

```diff
@@ -84,8 +84,17 @@ func (es *setupEpoch) FinalView() (uint64, error) {
 // TargetDuration returns the desired real-world duration for this epoch, in seconds.
 // This target is specified by the FlowEpoch smart contract in the EpochSetup event
 // and used by the Cruise Control system to moderate the block rate.
+// In case the epoch has extensions, the target duration is calculated based on the last extension, by calculating how many
+// views were added by the extension and adding the proportional time to the target duration.
 func (es *setupEpoch) TargetDuration() (uint64, error) {
-	return es.setupEvent.TargetDuration, nil
+	if len(es.extensions) == 0 {
+		return es.setupEvent.TargetDuration, nil
+	} else {
+		lastExtension := es.extensions[len(es.extensions)-1]
+		viewsPerEpoch := float64(es.setupEvent.FinalView - es.setupEvent.FirstView + 1)
+		extraViews := float64(lastExtension.FinalView - es.setupEvent.FinalView)
+		return es.setupEvent.TargetDuration + uint64(extraViews*float64(es.setupEvent.TargetDuration)/viewsPerEpoch), nil
```

Comment:
```suggestion
		viewDuration := float64(es.setupEvent.TargetDuration) / float64(es.setupEvent.FinalView-es.setupEvent.FirstView+1)
		lastExtension := es.extensions[len(es.extensions)-1]
		return es.setupEvent.TargetDuration + uint64(float64(lastExtension.FinalView-es.setupEvent.FinalView)*viewDuration)
```

Suggesting to compute this in terms of `viewDuration`, so the math is more similar to `TargetEndTime`.

Reply:
Yeah, we could do that as well. Don't mind to.